core/domain/expression_parser_test.py

```diff
@@ -0,0 +1,231 @@
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Unit tests for expression_parser.py."""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+from core.domain import expression_parser
+from core.tests import test_utils
+
+
+class HelperFunctionsUnitTests(test_utils.GenericTestBase):
+    """Test the 'contains_balanced_brackets' and 'is_algebraic' helper
+    functions.
+    """
+
+    def test_contains_balanced_brackets(self):
+        """Tests for contains_balanced_brackets method."""
+        self.assertTrue(expression_parser.contains_balanced_brackets(''))
+        self.assertTrue(expression_parser.contains_balanced_brackets('a+2'))
+        self.assertTrue(expression_parser.contains_balanced_brackets('(a / 2)'))
+        self.assertTrue(expression_parser.contains_balanced_brackets('[a/ 2]'))
+        self.assertTrue(expression_parser.contains_balanced_brackets(' {a/2} '))
+        self.assertTrue(expression_parser.contains_balanced_brackets('([a]/2)'))
+        self.assertTrue(expression_parser.contains_balanced_brackets(
+            '[(a/{ 2 })]'))
+        self.assertTrue(expression_parser.contains_balanced_brackets(
+            '(([{}]{})( ){[ ]})'))
+        self.assertTrue(expression_parser.contains_balanced_brackets(
+            '[[ [((()))[[[[[]{}]]{}]]()]] ]'))
+        self.assertTrue(expression_parser.contains_balanced_brackets(
+            '{( 2x^2 ) ^ [ 3/2 ]} / 4'))
+
+        self.assertFalse(expression_parser.contains_balanced_brackets('(a/2'))
+        self.assertFalse(expression_parser.contains_balanced_brackets('a/2]'))
+        self.assertFalse(expression_parser.contains_balanced_brackets('[)(]'))
+        self.assertFalse(expression_parser.contains_balanced_brackets('{ [} ]'))
+        self.assertFalse(expression_parser.contains_balanced_brackets(']]][[['))
+        self.assertFalse(expression_parser.contains_balanced_brackets(')({})'))
+        self.assertFalse(expression_parser.contains_balanced_brackets('4/{0/]'))
+        self.assertFalse(expression_parser.contains_balanced_brackets('(a/2]'))
+
+    def test_is_algebraic(self):
+        """Tests for is_algebraic method."""
+        self.assertTrue(expression_parser.is_algebraic('a^2.3'))
+        self.assertTrue(expression_parser.is_algebraic('abs(alpha)'))
+        self.assertTrue(expression_parser.is_algebraic('alpha/gamma'))
+        self.assertTrue(expression_parser.is_algebraic('A + 2/3'))
+
+        self.assertFalse(expression_parser.is_algebraic('1 + 2'))
```

Comment:
Something like 'alpha + zeta' is also not algebraic, right? Maybe check that? Basically some that have a valid and invalid identifier.

Reply:
I've extended the greek letters list so 'zeta' is now a valid identifier. I've added a test case with valid + invalid identifiers below.

Reply:
Test added.